controller_manager/test/mode_switch_test.cpp

```diff
@@ -0,0 +1,380 @@
+    virtual std::vector<std::string> getDeclaredClasses() {
+        std::vector<std::string> v;
+        for(std::map<std::string, std::string>::iterator it = classes.begin(); it != classes.end(); ++it){
+            v.push_back(it->first);
+        }
+        return v;
+    }
+    virtual void reload() {}
+};
+
+void update(controller_manager::ControllerManager &cm, const ros::TimerEvent& e){
+    cm.update(e.current_real, e.current_real - e.last_real);
+}
+
+class GuardROS{
+public:
+    ~GuardROS(){
+        ros::shutdown();
+        ros::waitForShutdown();
+    }
+};
+
+TEST(SwitchInterfacesTest, SwitchInterfaces)
+{
+    GuardROS guard;
+
+    SwitchBot bot;
+    ros::NodeHandle nh;
+
+    controller_manager::ControllerManager cm(&bot);
+
+    cm.registerControllerLoader(boost::make_shared<DummyControllerLoader>());
+
+    ros::Timer timer = nh.createTimer(ros::Duration(0.01), boost::bind(update, boost::ref(cm), _1));
+
+    ASSERT_TRUE(cm.loadController("group_pos"));
+    ASSERT_TRUE(cm.loadController("another_group_pos"));
+    ASSERT_TRUE(cm.loadController("group_vel"));
+    ASSERT_TRUE(cm.loadController("group_eff"));
+    ASSERT_TRUE(cm.loadController("single_pos"));
+    ASSERT_TRUE(cm.loadController("single_eff"));
+    ASSERT_TRUE(cm.loadController("invalid_group_pos"));
+    ASSERT_FALSE(cm.loadController("totally_random_name"));
+
+    {   // test mode conflict
+        std::vector<std::string> start, stop;
+        start.push_back("invalid_group_pos");
+        ASSERT_FALSE(cm.switchController(start,stop, controller_manager_msgs::SwitchControllerRequest::STRICT));
+    }
+    {   // test resource conflict
+        std::vector<std::string> start, stop;
+        start.push_back("group_pos");
+        start.push_back("group_vel");
+        ASSERT_FALSE(cm.switchController(start,stop, controller_manager_msgs::SwitchControllerRequest::STRICT));
+    }
+    {   // test pos group
+        std::vector<std::string> start, stop;
+        start.push_back("group_pos");
+        ASSERT_TRUE(cm.switchController(start,stop, controller_manager_msgs::SwitchControllerRequest::STRICT));
+        ASSERT_TRUE(cm.switchController(stop,start, controller_manager_msgs::SwitchControllerRequest::STRICT)); // clean-up
+    }
+    {   // test same mode switch
+        std::vector<std::string> start, stop, next_start;
+        start.push_back("group_pos");
+        ASSERT_TRUE(cm.switchController(start,stop, controller_manager_msgs::SwitchControllerRequest::STRICT));
+
+        next_start.push_back("group_pos");
+        ASSERT_TRUE(cm.switchController(next_start, start, controller_manager_msgs::SwitchControllerRequest::STRICT));
+
+        ASSERT_TRUE(cm.switchController(stop,next_start, controller_manager_msgs::SwitchControllerRequest::STRICT)); // clean-up
+    }
+    {   // test same mode switch
```

Comment:
Duplicated and wrong.. it should start "another_group_pos" to test the same HWI with a different controller